blocks/subquery/edit.tsx

```diff
@@ -0,0 +1,273 @@
+/* eslint-disable camelcase */
+import { useDebounce } from '@uidotdev/usehooks';
+import apiFetch from '@wordpress/api-fetch';
+import { InnerBlocks, useBlockProps } from '@wordpress/block-editor';
+import { useSelect } from '@wordpress/data';
+import {
+  useEffect,
+} from '@wordpress/element';
+import { addQueryArgs } from '@wordpress/url';
+
+import { Template } from '@wordpress/blocks';
+import { v4 as uuid } from '@lukeed/uuid';
+
+import type {
+  EditProps,
+  Option,
+} from '../query/types';
+
+import {
+  mainDedupe,
+} from '../../services/deduplicate';
+
+import buildTermQueryArgs from '../../services/buildTermQueryArgs';
+
+import QueryControls from '../../components/QueryControls';
+import './index.scss';
+
+interface PostTypeOrTerm {
+  name: string;
+  slug: string;
+}
+
+interface Window {
+  wpCurateQueryBlock: {
+    allowedPostTypes: PostTypeOrTerm[];
+    allowedTaxonomies: PostTypeOrTerm[];
+    parselyAvailable: string,
+  };
+}
+
+/**
+ * The edit function describes the structure of your block in the context of the
+ * editor. This represents what the editor will render when the block is used.
+ *
+ * @see https://developer.wordpress.org/block-editor/reference-guides/block-api/block-edit-save/#edit
+ *
+ * @return {WPElement} Element to render.
+ */
+export default function Edit({
+  attributes: {
+    backfillPosts = [],
+    deduplication = 'inherit',
+    maxNumberOfPosts = 10,
+    minNumberOfPosts = 1,
+    numberOfPosts = 5,
+    offset = 0,
+    posts: manualPosts = [],
+    postTypes = [],
+    searchTerm = '',
+    terms = {},
+    termRelations = {},
+    taxRelation = 'AND',
+    orderby = 'date',
+    uniqueId = '',
+  },
+  setAttributes,
+  context: {
+    postId,
+    query: {
+      include = '',
+    } = {},
+  },
+}: EditProps) {
+  const queryInclude = include.split(',').map((id: string) => parseInt(id, 10));
+  const index = queryInclude.findIndex((id: number) => id === postId);
+  const backfillPostsString = backfillPosts.join(',');
+
+  const {
+    wpCurateQueryBlock: {
+      allowedPostTypes = [],
+      allowedTaxonomies = [],
+      parselyAvailable = 'false',
+    } = {},
+  } = (window as any as Window);
+
+  if (!postTypes.length) {
+    setAttributes({ postTypes: allowedPostTypes });
+  }
+
+  // @ts-ignore
+  const [isPostDeduplicating, postTypeObject] = useSelect(
+    (select) => {
+      // @ts-ignore
+      const editor = select('core/editor');
+
+      // @ts-ignore
+      const type = editor.getEditedPostAttribute('type');
+      // @ts-ignore
+      const meta = editor.getEditedPostAttribute('meta');
+
+      return [
+        // It's possible for usePostMetaValue() to run here before useEntityProp() is available.
+        Boolean(meta?.wp_curate_deduplication),
+        // @ts-ignore
+        type ? select('core').getPostType(type) : null,
```

Comment:
imo we should move anything with a `@ts-ignore` into the body of the function and ensure that what we're returning is of a stable type. The line above forces the first parameter to be a boolean, but here, it could be a post type object, null, or whatever `.getPostType` returns in a failure case (maybe undefined?) - I'd recommend moving the logic that gets the post type object before the `return`, validating that it is in fact a post type object, and only including it in the return value if it is actually a post type object, else return null. That way, even though this function is riddled with `@ts-ignore` because the Gutenberg project doesn't export types for `useSelect` that can be understood here, we can at least guarantee that what comes _out_ of this function is either a post type object or null for the second array member.